File: quack-java/src/main/java/com/koushikdutta/quack/polyfill/timeout/JobScheduler.java
package com.koushikdutta.quack.polyfill.timeout;

import com.koushikdutta.quack.JavaScriptObject;
import com.koushikdutta.quack.QuackContext;

import java.util.Hashtable;

public interface JobScheduler {
    JobCanceller schedule(int timeout, Runnable runnable);
    JobCanceller schedule(Runnable runnable);

    /**
     * Installs setTimeout and setImmediate.
     * Installs a job scheduler for Promises.
     * @param quackContext
     */
    static void install(QuackContext quackContext, JobScheduler scheduler) {
        JavaScriptObject global = quackContext.getGlobalObject();

        {
            Hashtable<Integer, JobCanceller> timeouts = new Hashtable<>();
            global.set("setTimeout", new SetTimeout() {
                int timeoutCount;
                @Override
                public long invoke(JavaScriptObject cb, int milliseconds, Object... params) {
                    int myTimeout = timeoutCount++;
                    JobCanceller canceller = scheduler.schedule(milliseconds, () -> {
                        timeouts.remove(myTimeout);
                        global.call(params);
                    });
                    timeouts.put(myTimeout, canceller);
                    return myTimeout;
                }
            });
            global.set("clearTimeout", (ClearTimeout) timeout -> ClearTimeout.clear(timeout, timeouts));
        }

        {
            Hashtable<Integer, JobCanceller> timeouts = new Hashtable<>();
            global.set("setImmediate", new SetImmediate() {
                int timeoutCount;
                @Override
                public long invoke(JavaScriptObject cb, Object... params) {
                    int myTimeout = timeoutCount++;
                    JobCanceller canceller = scheduler.schedule(() -> {
                        timeouts.remove(myTimeout);
                        global.call(params);
                    });
                    timeouts.put(myTimeout, canceller);
                    return myTimeout;
                }
            });
            global.set("clearImmediate", (ClearTimeout) timeout -> ClearTimeout.clear(timeout, timeouts));
        }

        {
            Hashtable<Integer, JobCanceller> timeouts = new Hashtable<>();
            global.set("setInterval", new SetTimeout() {
                int timeoutCount;
                @Override
                public long invoke(JavaScriptObject cb, int milliseconds, Object... params) {
                    int myTimeout = timeoutCount++;
                    Runnable schedule = new Runnable() {
                        @Override
                        public void run() {
                            JobCanceller canceller = scheduler.schedule(milliseconds, () -> {
                                // reschedule first because the interval may cancel itself after.
                                run();
                                global.call(params);
                            });
                            timeouts.put(myTimeout, canceller);
                        }
                    };

                    schedule.run();
                    return myTimeout;
                }
            });
            global.set("clearInterval", (ClearTimeout) timeout -> ClearTimeout.clear(timeout, timeouts));
        }

        quackContext.setJobExecutor(scheduler::schedule);
    }
}

File: quack-polyfill/src/main/java/com/koushikdutta/quack/polyfill/job/JobScheduler.kt
package com.koushikdutta.quack.polyfill.job

import com.koushikdutta.quack.JavaScriptObject
import com.koushikdutta.quack.polyfill.QuackEventLoop
import com.koushikdutta.scratch.event.Cancellable
import java.util.*

fun QuackEventLoop.installJobScheduler() {
    val loop = loop
    val ctx = quack
    val global = ctx.globalObject
    run {
        val timeouts = Hashtable<Any, Cancellable>()
        global["setTimeout"] = ctx.coerceJavaToJavaScript(SetTimeout::class.java, object : SetTimeout {
            var timeoutCount = 0
            override fun execute(cb: JavaScriptObject?, milliseconds: Int, vararg params: Any): Int {
                if (cb == null) return -1
                val myTimeout = timeoutCount++
                val canceller = loop.postDelayed(milliseconds.toLong()) {
                    timeouts.remove(myTimeout)
                    cb.call(*params)
                }
                timeouts[myTimeout] = canceller
                return myTimeout
            }
        })
        global.set("clearTimeout", ctx.coerceJavaToJavaScript(ClearTimeout::class.java, ClearTimeout { timeout -> ClearTimeout.clear(timeout, timeouts) }))
    }
    run {
        val timeouts = Hashtable<Any, Cancellable>()
        global["setImmediate"] = ctx.coerceJavaToJavaScript(SetImmediate::class.java, object : SetImmediate {
            var timeoutCount = 0
            override fun execute(cb: JavaScriptObject?, vararg params: Any): Int {
                if (cb == null) return -1
                val myTimeout = timeoutCount++
                val canceller = loop.postImmediate {
                    timeouts.remove(myTimeout)
                    cb.call(*params)
                }
                timeouts[myTimeout] = canceller
                return myTimeout
            }
        })
        global.set("clearImmediate", ctx.coerceJavaToJavaScript(ClearTimeout::class.java, ClearTimeout { timeout -> ClearTimeout.clear(timeout, timeouts) }))
    }
    run {
        val timeouts = Hashtable<Any, Cancellable>()
        global["setInterval"] = ctx.coerceJavaToJavaScript(SetTimeout::class.java, object : SetTimeout {
            var timeoutCount = 0
            override fun execute(cb: JavaScriptObject?, milliseconds: Int, vararg params: Any): Int {
                if (cb == null) return -1
                val myTimeout = timeoutCount++
                val schedule: Runnable = object : Runnable {
                    override fun run() {
                        val canceller = loop.postDelayed(milliseconds.toLong()) {
                            // reschedule first because the interval may cancel itself after.
                            run()
                            cb.call(*params)
                        }
                        timeouts[myTimeout] = canceller
                    }
                }
                schedule.run()
                return myTimeout
            }
        })
        global.set("clearInterval", ctx.coerceJavaToJavaScript(ClearTimeout::class.java, ClearTimeout { timeout -> ClearTimeout.clear(timeout, timeouts) }))
    }
    ctx.setJobExecutor { runnable: Runnable ->
        loop.post {
            runnable.run()
        }
    }
}
